File: ai/services/file_upload.py
"""
File upload service for Python backend using Supabase Storage
Handles file processing and metadata storage after upload
"""

import os
import uuid
import logging
from typing import Optional, Dict, Any, List
from datetime import datetime
import mimetypes

from supabase import create_client, Client
from services.supabase_client import SupabaseClient

logger = logging.getLogger(__name__)

class FileUploadService:
    """
    Service for handling file uploads and processing with Supabase Storage
    """
    
    def __init__(self):
        """Initialize the file upload service"""
        self.supabase_client = SupabaseClient(use_secret_key=True)
        self.bucket_name = 'documents'
        
    async def process_uploaded_file(
        self, 
        file_path: str, 
        file_name: str, 
        file_url: str,
        user_id: str, 
        project_id: str,
        file_size: int,
        mime_type: str
    ) -> Dict[str, Any]:
        """
        Process an uploaded file and create database records
        
        Args:
            file_path: Path to file in Supabase Storage
            file_name: Original filename
            file_url: Public URL of the uploaded file
            user_id: ID of the user who uploaded the file
            project_id: ID of the project
            file_size: Size of the file in bytes
            mime_type: MIME type of the file
        
        Returns:
            Dictionary with processing results
        """
        try:
            # Create source record
            source_data = {
                'id': str(uuid.uuid4()),
                'title': file_name,
                'type': self._get_source_type(mime_type),
                'url': file_url,
                'content': '',  # Will be populated after text extraction
                'user_id': user_id,
                'metadata': {
                    'original_filename': file_name,
                    'upload_timestamp': datetime.utcnow().isoformat(),
                    'processing_status': 'pending',
                    'file_path': file_path,
                    'file_size': file_size,
                    'mime_type': mime_type
                },
                'created_at': datetime.utcnow().isoformat(),
                'updated_at': datetime.utcnow().isoformat()
            }
            
            # Add project_id if provided
            if project_id:
                source_data['project_id'] = project_id
            
            # Add file-specific fields if the schema supports them
            try:
                # Try to include file-specific columns (but not status for now)
                source_data.update({
                    'file_path': file_path,
                    'file_size': file_size,
                    'mime_type': mime_type
                })
            except:
                # If file-specific columns don't exist, store in metadata only
                pass
            
            # Insert source record
            result = self.supabase_client.insert('sources', source_data)
            
            if not result:
                raise Exception("Failed to create source record")
            
            # Start async processing for text extraction and embedding
            # This would typically be done in a background task
            await self._schedule_file_processing(source_data['id'], file_url, mime_type)
            
            return {
                'success': True,
                'source_id': source_data['id'],
                'message': 'File uploaded and processing started'
            }
            
        except Exception as e:
            logger.error(f"Error processing uploaded file: {e}")
            return {
                'success': False,
                'error': str(e)
            }
    
    async def _schedule_file_processing(self, source_id: str, file_url: str, mime_type: str):
        """
        Schedule background processing for the uploaded file
        This would typically use a task queue like Celery
        """
        try:
            # For now, we'll just update the status
            # In a production system, this would trigger background processing
            self.supabase_client.update(
                'sources',
                {'id': source_id},
                {
                    'metadata': {
                        'processing_status': 'scheduled',
                        'scheduled_at': datetime.utcnow().isoformat()
                    },
                    'updated_at': datetime.utcnow().isoformat()
                }
            )
            
            logger.info(f"Scheduled processing for source {source_id}")
            
        except Exception as e:
            logger.error(f"Error scheduling file processing: {e}")
    
    def _get_source_type(self, mime_type: str) -> str:
        """
        Determine source type based on MIME type
        """
        type_mapping = {
            'application/pdf': 'pdf',
            'text/plain': 'text',
            'text/markdown': 'markdown',
            'application/msword': 'document',
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document': 'document',
            'text/csv': 'spreadsheet',
            'application/json': 'data',
            'text/html': 'webpage',
            'application/rtf': 'document'
        }
        
        return type_mapping.get(mime_type, 'unknown')
    
    def delete_file(self, file_path: str) -> Dict[str, Any]:
        """
        Delete a file from Supabase Storage and clean up database records
        
        Args:
            file_path: Path to file in Supabase Storage
        
        Returns:
            Dictionary with deletion results
        """
        try:
            # Get source record first
            sources = self.supabase_client.select(
                'sources',
                filters={'file_path': file_path}
            )
            
            if not sources:
                return {'success': False, 'error': 'Source not found'}
            
            source = sources[0]
            source_id = source['id']
            
            # Delete from Supabase Storage
            # Note: This would typically be done via the storage API
            # For now, we'll just mark as deleted in the database
            
            # Delete embeddings
            self.supabase_client.delete('document_embeddings', {'source_id': source_id})
            
            # Delete source record
            self.supabase_client.delete('sources', {'id': source_id})
            
            logger.info(f"Deleted file and records for source {source_id}")
            
            return {
                'success': True,
                'message': 'File and associated records deleted successfully'
            }
            
        except Exception as e:
            logger.error(f"Error deleting file: {e}")
            return {
                'success': False,
                'error': str(e)
            }
    
    def get_file_info(self, source_id: str) -> Dict[str, Any]:
        """
        Get file information from the database
        
        Args:
            source_id: ID of the source
        
        Returns:
            Dictionary with file information
        """
        try:
            sources = self.supabase_client.select(
                'sources',
                filters={'id': source_id}
            )
            
            if not sources:
                return {'success': False, 'error': 'Source not found'}
            
            source = sources[0]
            
            return {
                'success': True,
                'file_info': {
                    'id': source['id'],
                    'title': source['title'],
                    'type': source['type'],
                    'url': source['url'],
                    'file_path': source['file_path'],
                    'file_size': source['file_size'],
                    'mime_type': source['mime_type'],
                    'status': source['status'],
                    'metadata': source['metadata'],
                    'created_at': source['created_at'],
                    'updated_at': source['updated_at']
                }
            }
            
        except Exception as e:
            logger.error(f"Error getting file info: {e}")
            return {
                'success': False,
                'error': str(e)
            }
    
    def list_project_files(self, user_id: str, project_id: str) -> Dict[str, Any]:
        """
        List all files for a project
        
        Args:
            user_id: ID of the user
            project_id: ID of the project
        
        Returns:
            Dictionary with file list
        """
        try:
            sources = self.supabase_client.select(
                'sources',
                filters={
                    'user_id': user_id,
                    'project_id': project_id
                },
                order_by='created_at.desc'
            )
            
            return {
                'success': True,
                'files': sources
            }
            
        except Exception as e:
            logger.error(f"Error listing project files: {e}")
            return {
                'success': False,
                'error': str(e)
            }

# Create singleton instance
file_upload_service = FileUploadService()

File: lib/file-upload.ts
/**
 * File upload service using Supabase Storage
 * Handles file uploads to Supabase buckets with proper error handling
 */

import { createClientComponentClient } from '@supabase/auth-helpers-nextjs'
import { v4 as uuidv4 } from 'uuid'

export interface UploadResult {
  success: boolean
  fileUrl?: string
  fileName?: string
  filePath?: string
  error?: string
  fileSize?: number
  mimeType?: string
}

export interface UploadProgress {
  loaded: number
  total: number
  percentage: number
}

export class FileUploadService {
  private supabase
  private bucketName = 'filedoc'

  constructor() {
    this.supabase = createClientComponentClient()
  }

  /**
   * Upload a file to Supabase Storage
   */
  async uploadFile(
    file: File,
    projectId: string,
    userId: string,
    onProgress?: (progress: UploadProgress) => void
  ): Promise<UploadResult> {
    try {
      // Validate file
      const validation = this.validateFile(file)
      if (!validation.valid) {
        return {
          success: false,
          error: validation.error
        }
      }

      // Generate unique file path
      const fileExtension = file.name.split('.').pop()
      const uniqueFileName = `${uuidv4()}.${fileExtension}`
      const filePath = `${userId}/${projectId}/${uniqueFileName}`

      // Check if bucket exists, create if not
      await this.ensureBucketExists()

      // Upload file with progress tracking
      const { data, error } = await this.supabase.storage
        .from(this.bucketName)
        .upload(filePath, file, {
          cacheControl: '3600',
          upsert: false
        })

      if (error) {
        console.error('Upload error:', error)
        return {
          success: false,
          error: `Upload failed: ${error.message}`
        }
      }

      // Get public URL
      const { data: publicUrlData } = this.supabase.storage
        .from(this.bucketName)
        .getPublicUrl(filePath)

      return {
        success: true,
        fileUrl: publicUrlData.publicUrl,
        fileName: file.name,
        filePath: filePath,
        fileSize: file.size,
        mimeType: file.type
      }

    } catch (error) {
      console.error('File upload error:', error)
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown upload error'
      }
    }
  }

  /**
   * Delete a file from Supabase Storage
   */
  async deleteFile(filePath: string): Promise<{ success: boolean; error?: string }> {
    try {
      const { error } = await this.supabase.storage
        .from(this.bucketName)
        .remove([filePath])

      if (error) {
        return {
          success: false,
          error: `Delete failed: ${error.message}`
        }
      }

      return { success: true }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown delete error'
      }
    }
  }

  /**
   * Download a file from Supabase Storage
   */
  async downloadFile(filePath: string): Promise<{ success: boolean; blob?: Blob; error?: string }> {
    try {
      const { data, error } = await this.supabase.storage
        .from(this.bucketName)
        .download(filePath)

      if (error) {
        return {
          success: false,
          error: `Download failed: ${error.message}`
        }
      }

      return {
        success: true,
        blob: data
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown download error'
      }
    }
  }

  /**
   * Get file info from Supabase Storage
   */
  async getFileInfo(filePath: string): Promise<{ success: boolean; info?: any; error?: string }> {
    try {
      const { data, error } = await this.supabase.storage
        .from(this.bucketName)
        .list(filePath.substring(0, filePath.lastIndexOf('/')), {
          limit: 1,
          search: filePath.substring(filePath.lastIndexOf('/') + 1)
        })

      if (error) {
        return {
          success: false,
          error: `Info retrieval failed: ${error.message}`
        }
      }

      return {
        success: true,
        info: data?.[0]
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown info retrieval error'
      }
    }
  }

  /**
   * Validate file before upload
   */
  private validateFile(file: File): { valid: boolean; error?: string } {
    // Check file size (50MB limit)
    const maxSize = 50 * 1024 * 1024 // 50MB
    if (file.size > maxSize) {
      return {
        valid: false,
        error: 'File size exceeds 50MB limit'
      }
    }

    // Check file type
    const allowedTypes = [
      'application/pdf',
      'text/plain',
      'text/markdown',
      'application/msword',
      'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
      'text/csv',
      'application/json',
      'text/html',
      'application/rtf'
    ]

    if (!allowedTypes.includes(file.type)) {
      return {
        valid: false,
        error: 'File type not supported. Please upload PDF, TXT, MD, DOC, DOCX, CSV, JSON, HTML, or RTF files.'
      }
    }

    return { valid: true }
  }

  /**
   * Ensure the documents bucket exists
   */
  private async ensureBucketExists(): Promise<void> {
    try {
      const { data: buckets } = await this.supabase.storage.listBuckets()
      console.log('Available buckets:', buckets)
      const bucketExists = buckets?.some(bucket => bucket.name === this.bucketName)

      if (!bucketExists) {
        console.log(`Bucket "${this.bucketName}" does not exist, creating...`)
        const { error } = await this.supabase.storage.createBucket(this.bucketName, {
          public: true,
          allowedMimeTypes: [
            'application/pdf',
            'text/plain',
            'text/markdown',
            'application/msword',
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            'text/csv',
            'application/json',
            'text/html',
            'application/rtf'
          ],
          fileSizeLimit: 52428800 // 50MB
        })

        if (error) {
          console.error('Failed to create bucket:', error)
        }
      }
    } catch (error) {
      console.error('Error checking/creating bucket:', error)
    }
  }

  /**
   * List files in a project directory
   */
  async listProjectFiles(userId: string, projectId: string): Promise<{ success: boolean; files?: any[]; error?: string }> {
    try {
      const { data, error } = await this.supabase.storage
        .from(this.bucketName)
        .list(`${userId}/${projectId}`)

      if (error) {
        return {
          success: false,
          error: `Failed to list files: ${error.message}`
        }
      }

      return {
        success: true,
        files: data
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown list error'
      }
    }
  }
}

// Create singleton instance
export const fileUploadService = new FileUploadService()
